Honor debug stop from configuration providers

alvsan09 opened the pull request:
#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Fixes: #10998 

Debug configuration providers can indicate to stop debugging by
returning `undefined` or `null` on either of the functions
- `resolveDebugConfiguration` or
- `resolveDebugConfigurationWithSubstitutedVariables`

The vscode API
https://code.visualstudio.com/api/references/vscode-api#DebugConfigurationProvider
specifies different handling for `undefined` and `null` as follows:

"Returning the value 'undefined' prevents the debug session from
starting. Returning the value 'null' prevents the debug session from
starting and opens the underlying debug configuration instead."

This implementation focuses on passing the `undefined` or `null`
response from the extension to the debug session manager where the
indication from the extension is handled.

Signed-off-by: Alvaro Sanchez-Leon <alvaro.sanchez-leon@ericsson.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

1. Download the following extension and place it under your theia/plugins folder 
https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-resolve-null.vsix
2. Start theia 
3. Testing behavior for an extension returning `undefined`: 
From the command palette (Ctrl + shft + P) , type `debug `, note trailing space is needed.
Select configuration named `Create JavaScript Debug Terminal`
Verify that no debug session is started and that a single `JavaScript Debug Terminal` instance is created.
This verifies the fix for issue: #10998
6. Testing behavior for an extension returning `null`:
From the command palette (Ctrl + shft + P) , type `debug `, note trailing space is needed.
Select configuration named `Resolve config to null` (this configuration is provided by the test extension loaded in step 1)
Verify that no debug session is started and that the `launch.json` file is opened.
 
![config_resolve_undefined_null](https://user-images.githubusercontent.com/76971376/162072780-025f80a2-1401-4358-a14a-5ca6f7008f57.gif)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/plugin/src/theia.d.ts:
```diff
@@ -9691,7 +9698,11 @@ export module '@theia/plugin' {
+            folder: WorkspaceFolder | undefined,
+            debugConfiguration: DebugConfiguration,
+            token?: CancellationToken
+        ): ProviderResult<DebugConfiguration | undefined | null>;
```
@alvsan09 I don't believe this api needs to be update since it is already aligned with [vscode](https://github.com/microsoft/vscode/blob/a3d53f76790d8c88f835afe85225ff6442703409/src/vscode-dts/vscode.d.ts#L13655-L13699) (and doing so will likely mark our api as incompatible in the compatibility report). Rather, the [`ProviderResult`](https://github.com/microsoft/vscode/blob/a3d53f76790d8c88f835afe85225ff6442703409/src/vscode-dts/vscode.d.ts#L2135) should be updated to accept `null` like in vscode:

```ts
export type ProviderResult<T> = T | undefined | null | Thenable<T | undefined | null>;
```

alvsan09 replied:
Nice catch Vnce! Thanks!
It is addressed in today's commit.

vince-fugnitto commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -394,6 +406,9 @@ export class DebugExtImpl implements DebugExt {
         return result;
     }
 
+    /**
+     * @deprecated since 1.25.0. Use $registerDebugConfigurationProvider with $resolveDebugConfigurationByHandle instead.
+     */
     async $resolveDebugConfigurations(debugConfiguration: theia.DebugConfiguration, workspaceFolderUri: string | undefined): Promise<theia.DebugConfiguration | undefined> {
```
Given how the method is pretty internal to the plugin-system (and not easily extended), it would be fine to simply break and not deprecate.

vince-fugnitto commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -419,6 +434,9 @@ export class DebugExtImpl implements DebugExt {
         return current;
     }
 
+    /**
+     * @deprecated since 1.25.0. Use $registerDebugConfigurationProvider with $resolveDebugConfigurationWithSubstitutedVariablesByHandle instead.
+     */
     async $resolveDebugConfigurationWithSubstitutedVariables(debugConfiguration: theia.DebugConfiguration, workspaceFolderUri: string | undefined):
```
Same as the comment for `$resolveDebugConfigurations`.

colin-grant-work reviewed (approved):
The code looks good and the provided tests work. @vince-fugnitto mentioned VSCode's `ProviderResult` type, and I think it might make sense to introduce something similar in our `packages/core/src/common/types.ts` file to reduce the number of times we have to write `| undefined | null`?

alvsan09 commented:
> I think it might make sense to introduce something similar in our `packages/core/src/common/types.ts` file to reduce the number of times we have to write `| undefined | null`?
Hi @colin-grant-work, 
I introduced a new type under `common` so the result can be passed across all the way to the browser, 

The new type is under a new commit so we can discuss this particular update separately!

I have updated all the related returns accordingly,  let me know what you think !

vince-fugnitto commented on packages/core/src/common/types.ts:
```diff
@@ -130,3 +130,5 @@ export namespace ArrayUtils {
     throw new Error(message);
 }
+
+export type ResultPromise<T> = Promise<T | undefined | null>;
```
@alvsan09 I would probably keep the same name `ProviderResult` unless you have a strong reason to deviate? I don't think `ResultPromise` would be a good name for this type.

paul-marechal replied:
I don't think we benefit from this type at all at them moment, I would remove it.

paul-marechal replied:
Specifically I don't think it's super useful to use for implementations, which is its main use right now.

Looking at VS Code they have a `ProviderResult` used in definitions, but they never mention it in implementations.

paul-marechal commented on packages/debug/src/browser/debug-session-manager.ts:
```diff
@@ -221,17 +235,31 @@ export class DebugSessionManager {
-    protected async resolveConfiguration(options: Readonly<DebugSessionOptions>): Promise<InternalDebugSessionOptions> {
+    protected async resolveConfiguration(
+        options: Readonly<DebugSessionOptions>
+    ): ResultPromise<InternalDebugSessionOptions> {
```
Those methods are implementations. As long as the interface to implement defines return types as `ProviderResult<T>` which really means `T | undefined | null | Thenable<T | undefined  | null>` we can retype implementations to something more specific/convenient. The interface says "one of those can be returned" while the implementation will only return a subset of that.

It's fine to type as `Promise<InternalDebugSessionOptions | undefined>` here.

alvsan09 commented:
@colin-grant-work, @paul-marechal, @vince-fugnitto,
After offline discussions with Paul and Vince, there seems to be a preference to not introduce the new type,
so I have reset the tip of the branch to the previous commit (before the introduction of the new type), 

Colin had already approved that commit, @vince-fugnitto  and @paul-marechal any pending items before approving this PR?

alvsan09 commented:
Re-based to latest master

colin-grant-work commented:
My approval stands, with or without the additional type. 👍

vince-fugnitto reviewed (approved):
LGTM!
